hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java

```diff
@@ -41,131 +16,86 @@
+     *     2. If the call is not historical, the method initializes the contract call context with the current state
+     *     and proceeds to call the contract.
+     * </p>
+     *
+     * @param params the call service parameters
+     * @param ctx the contract call context
+     * @return {@link HederaEvmTransactionProcessingResult} of the contract call
+     * @throws MirrorEvmTransactionException if any pre-checks
+     * fail with {@link IllegalStateException} or {@link IllegalArgumentException}
      */
-    private Bytes estimateGas(final CallServiceParameters params) {
-        final var processingResult = doProcessCall(params, params.getGas(), true);
-        validateResult(processingResult, ETH_ESTIMATE_GAS);
-
-        final var gasUsedByInitialCall = processingResult.getGasUsed();
-
-        // sanity check ensuring gasUsed is always lower than the inputted one
-        if (gasUsedByInitialCall >= params.getGas()) {
-            return Bytes.ofUnsignedLong(gasUsedByInitialCall);
+    protected HederaEvmTransactionProcessingResult callContract(CallServiceParameters params,
+                                                                ContractCallContext ctx) throws MirrorEvmTransactionException {
+        // if we have historical call, then set the corresponding record file in the context
+        if (params.getBlock() != BlockType.LATEST) {
+            ctx.setRecordFile(recordFileService
+                    .findByBlockType(params.getBlock())
+                    .orElseThrow(BlockNumberNotFoundException::new));
         }
-
-        final var estimatedGas = binaryGasEstimator.search(
-                (totalGas, iterations) -> updateGasMetric(ETH_ESTIMATE_GAS, totalGas, iterations),
-                gas -> doProcessCall(params, gas, false),
-                gasUsedByInitialCall,
-                params.getGas());
-
-        return Bytes.ofUnsignedLong(estimatedGas);
+        // initializes the stack frame with the current state or historical state (if the call is historical)
+        ctx.initializeStackFrames(store.getStackedStateFrames());
+        return doProcessCall(params, params.getGas(), true, ctx);
     }
 
-    private HederaEvmTransactionProcessingResult doProcessCall(
-            CallServiceParameters params, long estimatedGas, boolean restoreGasToThrottleBucket) {
+    protected HederaEvmTransactionProcessingResult doProcessCall(CallServiceParameters params,
+                                                                 long estimatedGas,
+                                                                 boolean restoreGasToThrottleBucket,
+                                                                 ContractCallContext ctx) throws MirrorEvmTransactionException {
         try {
-            var result = mirrorEvmTxProcessor.execute(params, estimatedGas);
+            var result = mirrorEvmTxProcessor.execute(params, estimatedGas, params.getTracerType(), ctx);
```

Comment:
We already have the tracer type in the params:
```suggestion
            var result = mirrorEvmTxProcessor.execute(params, estimatedGas, ctx);
```